util/certs/certs.go

```diff
@@ -76,11 +77,33 @@ func DecodeCertPEM(encoded []byte) (*x509.Certificate, error) {
 
 // DecodePrivateKeyPEM attempts to return a decoded key or nil
 // if the encoded input does not contain a private key.
-func DecodePrivateKeyPEM(encoded []byte) (*rsa.PrivateKey, error) {
+func DecodePrivateKeyPEM(encoded []byte) (crypto.Signer, error) {
 	block, _ := pem.Decode(encoded)
 	if block == nil {
 		return nil, nil
 	}
 
-	return x509.ParsePKCS1PrivateKey(block.Bytes)
+	pkcs1Key, pkcs1Err := x509.ParsePKCS1PrivateKey(block.Bytes)
+	if pkcs1Err == nil {
+		return crypto.Signer(pkcs1Key), nil
+	}
+
+	// ParsePKCS1PrivateKey will fail with errors.New for many reasons
+	// including if the format is wrong, so we can retry with PKCS8 or EC
+	// https://golang.org/src/crypto/x509/pkcs1.go#L58
+	pkcs8Key, pkcs8Err := x509.ParsePKCS8PrivateKey(block.Bytes)
+	if pkcs8Err == nil {
+		pkcs8Signer, ok := pkcs8Key.(crypto.Signer)
+		if !ok {
+			return nil, errors.New("x509: certificate private key does not implement crypto.Signer")
+		}
+		return pkcs8Signer, nil
+	}
+
+	ecKey, ecErr := x509.ParseECPrivateKey(block.Bytes)
+	if ecErr == nil {
+		return crypto.Signer(ecKey), nil
+	}
+
+	return nil, pkcs1Err
```

Comment:
Might be good to collect all the errors from the various x590.Parse* functions and return the aggregate here instead, similar to: https://github.com/kubernetes-sigs/cluster-api/blob/2bb56ed2eb2a91cb59b740cb286992952da3dfed/controllers/cluster_controller.go#L177